<a href="https://colab.research.google.com/github/gitmystuff/DSChunks/blob/main/Bias_Variance_Example.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Bias Variance Example

### Final Model

In [ ]:
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# model = RandomForestClassifier(random_state=random_state).set_params(**best_parameters) # * args, ** kwargs
# model.fit(X_train[features_to_model], y_train)
# predictions = model.predict(X_test[features_to_model])
# print('accuracy:', accuracy_score(predictions, y_test))
# print(confusion_matrix(y_test, predictions))
# print(classification_report(y_test, predictions))

accuracy: 0.8866666666666667
[[123  20]
 [ 14 143]]
              precision    recall  f1-score   support

         0.0       0.90      0.86      0.88       143
         1.0       0.88      0.91      0.89       157

    accuracy                           0.89       300
   macro avg       0.89      0.89      0.89       300
weighted avg       0.89      0.89      0.89       300



<pre>
accuracy: 0.8866666666666667
[[123  20]
 [ 14 143]]
              precision    recall  f1-score   support

         0.0       0.90      0.86      0.88       143
         1.0       0.88      0.91      0.89       157

    accuracy                           0.89       300
   macro avg       0.89      0.89      0.89       300
weighted avg       0.89      0.89      0.89       300
</pre>

## Bias Variance

In [ ]:
# from mlxtend.evaluate import bias_variance_decomp

# avg_expected_loss, avg_bias, avg_var = bias_variance_decomp(
#     model,
#     X_train[features_to_model].values, # Convert X_train to NumPy array
#     y_train.values, # Convert y_train to NumPy array
#     X_test[features_to_model].values, # Convert X_test to NumPy array
#     y_test.values, # Convert y_test to NumPy array
#     loss='0-1_loss',
#     random_seed=123)

# print('Average expected loss: %.3f' % avg_expected_loss)
# print('Average bias: %.3f' % avg_bias)
# print('Average variance: %.3f' % avg_var)

# Average expected loss: 0.117
# Average bias: 0.113
# Average variance: 0.023

Average expected loss: 0.117
Average bias: 0.113
Average variance: 0.023


## Bias Variance Terms

**Bias**

*   **Definition:** Bias is the error introduced by approximating a real-world problem, which may be extremely complicated, by a much simpler model.
*   **Explanation:** A high-bias model oversimplifies the relationship between the features and the target variable, leading to underfitting. In your code, an average bias of 0.080 suggests that the model might be oversimplifying the relationship between the selected features ('features\_to\_model') and the likelihood of surviving the Titanic.

**Variance**

*   **Definition:** Variance is the error introduced by the model's sensitivity to fluctuations in the training data.
*   **Explanation:** A high-variance model captures the noise and outliers in the training data, leading to overfitting. In your code, an average variance of 0.034 indicates that the model is somewhat sensitive to the specific training data used. It might be capturing some random fluctuations that don't generalize well to new data.

**Expected Loss**

*   **Definition:** The average loss (error) you expect to see when applying your model to new, unseen data.
*   **Explanation:** It's a combination of bias and variance. In your code, the average expected loss of 0.087 represents the overall error rate you can expect from your model on new data.

**Interpretation**

The results suggest that your model has a slightly higher bias than variance. This indicates a tendency towards underfitting. You could try the following to improve model performance:

*   **Add more features:** Including more relevant features might help the model capture more complex relationships in the data.
*   **Use a more complex model:** Consider using a more flexible model that can capture non-linear patterns.
*   **Reduce regularization:** If you're using regularization, reducing the penalty might allow the model to learn more from the data.

## Bias Variance Relationship

Here's a breakdown of the relationship between bias, variance, and the datasets:

**Bias**

* Primarily related to the **training** dataset.
* High bias means the model is **too simple** to capture the underlying patterns in the training data, leading to underfitting.
* Imagine trying to fit a straight line to a curved dataset - no matter how much data you have, the line won't capture the curve well. That's high bias.

**Variance**

* Primarily related to the **training** dataset as well.
* High variance means the model is **too complex** and fits the noise and outliers in the training data, leading to overfitting.
* Think of trying to fit a highly flexible curve to a dataset with some random scatter. The curve might hit every point in the training set, including the random noise, but it won't generalize well to new data. That's high variance.

**Testing Dataset**

* The testing dataset is used to **evaluate** the model's performance on unseen data.
* It helps to estimate how well the model generalizes to new data and identify potential issues like overfitting (high variance).

**Computing the Relationship**

The `bias_variance_decomp` function you provided uses a technique called **bootstrap sampling**. Here's a simplified explanation of what's going on behind the scenes:

1.  **Resampling:** It creates multiple subsets of the training data by randomly sampling with replacement.
2.  **Model Training:** It trains a separate model on each subset.
3.  **Prediction:** It uses each trained model to make predictions on the testing data.
4.  **Bias Calculation:** It calculates the average difference between the predictions of all the models and the true values in the testing data. This reflects the model's tendency to underfit (high bias) or overfit (low bias).
5.  **Variance Calculation:** It calculates the variability of predictions across different models for each data point in the testing set. High variability indicates high variance (overfitting).

**Key Takeaway**

Both bias and variance are primarily related to the training data and how well the model learns from it. The testing data is used to evaluate the combined effect of bias and variance on the model's ability to generalize.

## Bias

In the context of the `bias_variance_decomp` function and the provided code, an average bias of **0.113** means that your model has a relatively **low bias**. Bias refers to the error introduced by approximating a real-world problem by a simpler model. A low bias suggests that your model is **not oversimplifying** the relationship between the features and the target variable. [cite: 140, 141, 142, 143]

The average bias in the context of the `bias_variance_decomp` function, typically ranges from 0 to 1. This is because the function calculates the bias using the 0-1 loss, which measures the fraction of incorrect predictions.

*   A lower average bias (closer to 0) indicates better performance, as it suggests the model is making fewer errors on average.
*   A higher average bias (closer to 1) indicates poorer performance, with the model making more errors on average.

**Good Bias (Low Bias)**

*   An average bias of 0.113, like in your example, suggests a good bias. This means the model is making correct predictions about 88.7% of the time on average.
*   Other examples of good bias values might be 0.05, 0.10, or 0.15, depending on the complexity of the problem and the expected accuracy.

**Bad Bias (High Bias)**

*   An average bias of 0.5 or higher would generally be considered a bad bias. This means the model is making incorrect predictions at least 50% of the time, which is often no better than random guessing for a binary classification problem.
*   Other examples of bad bias might be 0.4, 0.3, or 0.2, depending on the context.

**Important Considerations**

*   The acceptable range for average bias depends on the specific problem and dataset. Some problems might have inherently higher bias due to noise or limited features.
*   The goal is to minimize both bias and variance for optimal model performance.

## Variance

In the context of the `bias_variance_decomp` function and the 0-1 loss, an average variance of **0.023** means that your model has a relatively **low variance**.

Variance refers to the error introduced by the model's sensitivity to fluctuations in the training data. A low variance suggests that your model is **not overly sensitive** to the specific nuances or noise present in the training set, and thus, it's more likely to generalize well to unseen data.

Similar to average bias, the average variance typically ranges from 0 to 1 when using the `bias_variance_decomp` function with the 0-1 loss.

*   A lower average variance (closer to 0) indicates better performance, as it suggests the model is less sensitive to fluctuations in the training data and more likely to generalize well.
*   A higher average variance (closer to 1) indicates poorer performance, with the model potentially overfitting the training data and capturing noise that doesn't generalize.

**Good Variance (Low Variance)**

*   An average variance of 0.023, as in your case, is generally considered a good variance score. It suggests the model is relatively stable and not overly influenced by the specificities of the training data.
*   Other examples of good variance scores might be 0.01, 0.05, or 0.10, depending on the dataset and the model's complexity.

**Bad Variance (High Variance)**

*   An average variance exceeding 0.2 or 0.3 could be considered a bad variance score, especially for simpler datasets or models. This might indicate that the model is overfitting the training data and capturing noise that won't generalize to new data.
*   Other examples of bad variance scores might include 0.25, 0.4, or 0.5, depending on the context.

**Important Notes**

*   The acceptable range for average variance can vary depending on the specific problem, dataset, and model used.
*   The goal is to find a balance between bias and variance to achieve optimal model performance, minimizing both underfitting and overfitting.